File: web/src/api/externalServices.ts
import { api } from '@/lib/api';

export type ExternalServiceName = 'tmdb' | 'omdb' | 'tvdb';

export type ExternalServiceOutcome =
  | 'ok'
  | 'auth_failed'
  | 'network'
  | 'timeout'
  | 'proxy_failed'
  | 'dns_blocked';

export interface ExternalServiceDTO {
  service: ExternalServiceName;
  enabled: boolean;
  api_key_masked: string;
  api_key_configured: boolean;
  proxy_url_set: boolean;
  proxy_auth_set: boolean;
  proxy_scheme?: string;
  proxy_host?: string;
  last_test_at?: string;
  last_test_outcome?: ExternalServiceOutcome;
  last_test_message?: string;
  // Story 489 (B-17): runtime validation status. Empty when the service
  // was never validated. 'valid' = inline probe or POST /test succeeded.
  // 'invalid_key' = either a live 401 was reported by the TMDB client
  // OR a validate-on-save Upsert was rejected by the upstream.
  last_validation_at?: string;
  last_validation_status?: 'valid' | 'invalid_key';
  last_validation_message?: string;
}

export interface ExternalServiceUpsertRequest {
  enabled: boolean;
  // Pointer semantics: omit → unchanged, "" → clear, non-empty → set.
  api_key?: string;
  proxy_url?: string;
  proxy_username?: string;
  proxy_password?: string;
}

export interface ExternalServiceTestResponse {
  outcome: ExternalServiceOutcome;
  message?: string;
  latency_ms: number;
}

export async function listExternalServices(): Promise<ExternalServiceDTO[]> {
  const res = await api<{ services: ExternalServiceDTO[] }>('/external-services');
  return res.services;
}

export async function upsertExternalService(
  service: ExternalServiceName,
  body: ExternalServiceUpsertRequest,
): Promise<ExternalServiceDTO> {
  return api<ExternalServiceDTO>(`/external-services/${service}`, {
    method: 'PUT',
    body,
  });
}

export async function testExternalService(
  service: ExternalServiceName,
): Promise<ExternalServiceTestResponse> {
  return api<ExternalServiceTestResponse>(`/external-services/${service}/test`, {
    method: 'POST',
  });
}

File: internal/shared/http/dto/external_services.go
package dto

import "time"

// ExternalServiceDTO is the masked wire shape returned by GET / PUT.
// Plaintext keys NEVER appear here; api_key_masked is "****" + last4.
type ExternalServiceDTO struct {
	Service          string     `json:"service"            example:"tmdb"`
	Enabled          bool       `json:"enabled"`
	APIKeyMasked     string     `json:"api_key_masked"     example:"****abcd"`
	APIKeyConfigured bool       `json:"api_key_configured"`
	ProxyURLSet      bool       `json:"proxy_url_set"`
	ProxyAuthSet     bool       `json:"proxy_auth_set"`
	ProxyScheme      string     `json:"proxy_scheme,omitempty"  example:"socks5"`
	ProxyHost        string     `json:"proxy_host,omitempty"    example:"proxy.example.com:1080"`
	LastTestAt       *time.Time `json:"last_test_at,omitempty"`
	LastTestOutcome  string     `json:"last_test_outcome,omitempty" example:"ok"`
	LastTestMessage  string     `json:"last_test_message,omitempty"`
	// Story 489 (B-17): runtime validation status. Empty when the
	// service was never validated (live 401 hook never fired AND the
	// operator has not saved a key); "valid" or "invalid_key" otherwise.
	LastValidationAt      *time.Time `json:"last_validation_at,omitempty"`
	LastValidationStatus  string     `json:"last_validation_status,omitempty" enums:"valid,invalid_key"`
	LastValidationMessage string     `json:"last_validation_message,omitempty"`
}

// ExternalServiceListResponse wraps the list endpoint return.
type ExternalServiceListResponse struct {
	Services []ExternalServiceDTO `json:"services"`
}

// ExternalServiceUpsertRequest carries the operator-supplied PUT body.
// Pointer fields implement the PUT semantics: nil = unchanged in JSON
// (omitted by the client), explicit "" = clear, non-empty = set.
type ExternalServiceUpsertRequest struct {
	Enabled       bool    `json:"enabled"`
	APIKey        *string `json:"api_key,omitempty"`
	ProxyURL      *string `json:"proxy_url,omitempty"`
	ProxyUsername *string `json:"proxy_username,omitempty"`
	ProxyPassword *string `json:"proxy_password,omitempty"`
}

// ExternalServiceTestResponse is the result of POST :test.
type ExternalServiceTestResponse struct {
	Outcome   string `json:"outcome"      example:"ok"`
	Message   string `json:"message,omitempty"`
	LatencyMS int64  `json:"latency_ms"`
}
